File: codewars/solutions/javascript/4 kyu/Boggle Word Checker/solution.js
let STEPS = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]];

let isVisited = (arr, pos) => arr.filter(p => p[0] === pos[0] && p[1] === pos[1]).length !== 0;
let getValue = (arr, x, y) => {try {return arr[x][y];} catch (e) {return null;}};


let checkWord = (board, word) => {
    let queue = [];

    for (let [i, w] of board.entries())
        for (let [j, c] of w.entries())
            if (c === word[0])
                queue.push([word.slice(1), [i, j], [[i, j]]]);

    while (queue.length) {
        let [w, [x, y], visited] = queue.shift();

        if (w.length === 0)
            return true;

        STEPS
            .map(arr => [x + arr[0], y + arr[1]])
            .filter(pos => !isVisited(visited, pos) && getValue(board, ...pos) === w[0])
            .forEach(pos => queue.push([w.slice(1), pos, [...visited, pos]]));

    }

    return false;
};

File: codewars/solutions/python/4 kyu/Boggle Word Checker/solution.py
from heapq import heappop, heappush

STEPS = (-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1)


def find_word(board, word):
    queue = [
        (word[1:], (i, j), {(i, j)})
        for i, w in enumerate(board) for j, c in enumerate(w)
        if c == word[0]
    ]

    while queue:
        w, (x, y), visited = heappop(queue)

        if not w:
            return True

        for x_step, y_step in STEPS:
            xx, yy = pos = x + x_step, y + y_step

            if pos not in visited and len(board) > xx >= 0 and len(board[0]) > yy >= 0 and board[xx][yy] == w[0]:
                heappush(queue, (w[1:], pos, visited | {pos}))

    return False